.github/workflows/CommonCI.yml

```
@@ -0,0 +1,217 @@
+        if: ${{ inputs.julia_version == '1.10' || inputs.julia_version == 'lts' }}
+        env:
+          JULIA_PKG_SERVER_REGISTRY_PREFERENCE: eager
+
+      # Run the tests (pjrt or ifrt or both)
+      - name: "Setup Runtime Preferences (PJRT)"
+        if: ${{ inputs.runtime == 'pjrt' || inputs.runtime == 'both' }}
+        uses: "DamianReeves/write-file-action@master"
+        with:
+          path: "LocalPreferences.toml"
+          write-mode: "overwrite"
+          contents: |
+            [Reactant]
+            xla_runtime = "PJRT"
+      - name: "Run Tests (PJRT)"
+        if: ${{ inputs.runtime == 'pjrt' || inputs.runtime == 'both' }}
+        timeout-minutes: 120
+        run: |
+          import Pkg
+          Pkg.Registry.update()
+          Pkg.test(;
+            coverage="user",
+            allow_reresolve=parse(Bool, get(ENV, "ALLOW_RERESOLVE", "true"))
+          )
+        shell: julia --color=yes --code-coverage=user --depwarn=yes --project=. {0}
+        env:
+          ALLOW_RERESOLVE: ${{ !inputs.downgrade_testing }}
+          JULIA_PKG_SERVER_REGISTRY_PREFERENCE: eager
+          REACTANT_TEST_GROUP: ${{ inputs.test_group }}
+          XLA_FLAGS: "--xla_force_host_platform_device_count=12"
+          JULIA_DEBUG: "Reactant,Reactant_jll"
+
+      - name: "Setup Runtime Preferences (IFRT)"
+        if: ${{ inputs.runtime == 'ifrt' || inputs.runtime == 'both' }}
+        uses: "DamianReeves/write-file-action@master"
+        with:
+          path: "LocalPreferences.toml"
+          write-mode: "overwrite"
+          contents: |
+            [Reactant]
+            xla_runtime = "IFRT"
+      - name: "Run Tests (IFRT)"
+        if: ${{ inputs.runtime == 'ifrt' || inputs.runtime == 'both' }}
+        timeout-minutes: 120
+        run: |
+          import Pkg
+          Pkg.Registry.update()
+          Pkg.test(;
+            coverage="user",
+            allow_reresolve=parse(Bool, get(ENV, "ALLOW_RERESOLVE", "true"))
+          )
+        shell: julia --color=yes --code-coverage=user --depwarn=yes --project=. {0}
+        env:
+          ALLOW_RERESOLVE: ${{ !inputs.downgrade_testing }}
+          JULIA_PKG_SERVER_REGISTRY_PREFERENCE: eager
+          REACTANT_TEST_GROUP: ${{ inputs.test_group }}
+          XLA_FLAGS: "--xla_force_host_platform_device_count=12"
+          JULIA_DEBUG: "Reactant,Reactant_jll"
+
+      - name: "Upload MLIR modules"
+        uses: actions/upload-artifact@v4
+        timeout-minutes: 10
+        if: ${{ always() }}
+        with:
+          name: "mlir-${{ inputs.julia_version }}-${{ inputs.os }}-${{ inputs.runtime }}-assertions=${{ inputs.assertions }}-${{ github.event_name }}-test_group=${{ inputs.test_group }}-${{ inputs.localjll }}"
+          path: "**/*.mlir"
+          retention-days: 90
+          overwrite: false
+
+      - name: Save Julia depot cache on cancel or failure
+        if: ${{ inputs.localjll && (cancelled() || failure()) }}
```

I think it'd be worth to always do the upload in case of failure, not just when doing a local build (although with the upcoming cache restrictions we'll need to rethink our caching entirely):
```suggestion
        if: cancelled() || failure()
```